src/core/reactor.cc

```diff
@@ -4584,7 +4596,16 @@ void smp::configure(const smp_options& smp_opts, const reactor_options& reactor_
     _qs_owner = decltype(smp::_qs_owner){new smp_message_queue* [smp::count], qs_deleter{}};
     _qs = _qs_owner.get();
     for(unsigned i = 0; i < smp::count; i++) {
-        smp::_qs_owner[i] = reinterpret_cast<smp_message_queue*>(operator new[] (sizeof(smp_message_queue) * smp::count));
+        smp::_qs_owner[i] = reinterpret_cast<smp_message_queue*>(operator new[] (sizeof(smp_message_queue) * smp::count
+        // smp_message_queue has members with hefty alignment requirements.
+        // if we are reactor thread, or not running with dpdk, doing this
+        // new default aligned seemingly works, as does reordering
+        // dlinit dependencies (ugh). But we should enforce calling out to
+        // aligned_alloc, instead of pure malloc, if possible.
```

Comment:
i don't quite understand this part.. 

i am not questioning the correctness of this change. as `__STDCPP_DEFAULT_NEW_ALIGNMENT__` is 16. and `sizeof(smp_message_queue)` is greater than 16. so this change should enforce the alignment of the allocated storage.

but how is this related to if we are seastar reactor or using a dpdk reactor? do you mean, we could perform unaligned access when using dpdk?

Reply:
When building with dpdk, we get a slightly different init sequence, which, when fixing the original SIF such that non-reactor allocation goes to libc malloc from the get-go, causes the above new/alloc to end up in different memory blocks -> without SIF fix we get nice, page-level memory. With it, we end up in normal malloc and get 16/8-aligned block. I can only assume because in the former case, we've done a bunch of library allocs actually via seastar allocator, which in turn means we've stolen a bunch of pages and thus impact where malloc, once called, will place things. 

Now, I assume that normally the allocation here should compensate for this unless calling a new(aligned) alloc, and over-alloc + adjust pointers + make the end result object messy internally (finding the original memory etc), but either due to the raw memory alloc + alloc in place, or something else I am missing, this does not happen. 

Honestly, I would love to have a more exact and detailed explanation, but my tolerance for single-instructioning myself through libc etc only goes so far. Should also be observed that the failure does not happen if running in gdb either, possibly because some slight .so loading difference there as well. 

In any case, what I could observe is alignment of memory we actually get in the two variants, and when if works we get properly aligned memory, even though the new goes to a simple malloc. Luck?

Reply:
The seastar allocator has a power-of-two thing so it gets naturally aligned, and I guess libc does not, so it makes sense.

Reply:
And, the original code was just broken - it bypassed the new-expression to call operator new directly (maybe to avoid allocating the array size), but "forgot" to align properly.

Reply:
Note however that we could get past this (by luck?) even when running with such options that the new call was always served by glibc malloc. Otoh, even then, we'd mostly have smp_count > 1 (i.e. 8 or so), and thus at least fall into larger alloc domain.

Reply:
thank you very much Calle, all clear!